core/src/Kokkos_Swap.hpp

```diff
@@ -0,0 +1,67 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#ifndef KOKKOS_SWAP_HPP
+#define KOKKOS_SWAP_HPP
+
+#include <Kokkos_Macros.hpp>
+
+#include <cstddef>
+#include <type_traits>
+#include <utility>
+
+namespace Kokkos {
+
+template <class T>
+KOKKOS_FUNCTION constexpr std::enable_if_t<std::is_move_constructible_v<T> &&
+                                           std::is_move_assignable_v<T>>
+swap(T& a, T& b) noexcept(std::is_nothrow_move_constructible_v<T>&&
```

Comment:
i understand that it would be more appropriate to make it a constraint, but in this case we only have effectively only single overload , so using sfinea makes the error message very cryptic. can we jsut leave it as a static assert ?

Reply:
> Can we jsut leave it as a static assert ?

No we would not be able to detect if the underlying element type is swappable in the overload for arrays.